### DISCUSSION TITLE
In-app messaging implementation

### DISCUSSION THREAD

**Discussion by EvgeniyKrishtopa** [invertase/react-native-firebase#6567] [General]
I started implementation firebase in-app messaging and now have a few questions:

1. How can I retrieve data from message like message title, body, button text, image url and additional options and use it after getting message?
2. How can I handle events if I need to show message only on necessary screen, for example, home screen my app

**Comment by mikehardy** [comment]
Hi there -

1- currently you cannot, the in-app-messaging APIs have greatly expanded after the implementation here was completed long ago, and the new APIs, which allow registration of click listeners, with callbacks including message accessors, have not been implemented
2- this will be very very project-specific. You are in control via what events you generate + set as display conditions (https://firebase.google.com/docs/in-app-messaging/explore-use-cases#schedule-messages) and/or message display suppression via APIs here as a big on/off switch


If 1 is a need for your project, we could merge a PR if proposed - it would need to have javascript APIs that allow you to add react-native event listeners to our delegate implementation on iOS (https://firebase.google.com/docs/reference/ios/firebaseinappmessaging/api/reference/Protocols/FIRInAppMessagingDisplayDelegate) and register listeners on Android (https://firebase.google.com/docs/reference/android/com/google/firebase/inappmessaging/FirebaseInAppMessaging) - and add code in those native implementations to serve as EventEmitters to call those listeners. When the native code listeners are called they would need to parse out the message type / message contents to populate an object (with related typings in the typescript files) for return to the javascript layer where it would then be available to the app

There are several examples in the code of event emitters doing this sort of thing and the core/app package handles a lot of the event source / event listener registration stuff since it's a common pattern across firebase modules. The auth package might be a good one since the auth token changed / user changed pattern is similar - it's a whole app listener registration and the events can be completely native SDK generated

**Reply by EvgeniyKrishtopa** [reply]
Thanks for answers! 
2 - I set custom event in console https://prnt.sc/KYkCzmOVq0be and try to display message when I visit home screen via this logic https://prnt.sc/i6pCjiybIqNR
I am testing my campaign but behaviour weird and message appears on any screen on foreground event despite my logic. What I should change? Can you give me a link with example of proper implementation custom event and display conditions?

**Reply by mikehardy** [reply]
I think the react-native architecture ("the whole app lives in *one* native screen / MainActivity where we load a javascript bundle and load javascript-level screens) does not work well with the native-Android perspective of "apps will have multiple Ativities and we consider those screens". So perhaps you want to take advantage of suppression on all but the javascript-level main screen since from a native perspective it's *all* the main screen? Just a guess though. I do not have examples of any of this, I'm not sure anyone does, it appears to be a very lightly used module based on npmjs.com download counts so I haven't fielded many questions about it nor played with it myself

**Reply by shortSparrow** [reply]
As I understand this technology works with applications where one screen is one activity, does it mean that it will not work with single activity android app (which use fragment, and compose)?

**Reply by mikehardy** [reply]
@nizkiyVorobey  I'm not sure - it probably should be tested for your specific requirements.

However, by my read of the code it looks like events will only be triggered if the app is foreground, the campaign is not over it's rate limit for that in-app-message, and a *specific analytics event chosen during in-app-messaging campaign construction* is logged - reference https://firebase.google.com/docs/in-app-messaging/compose-campaign#schedule-your-message

So I think it is actually orthogonal to view construction style

Thus to focus my answer above to the question above, if the question is "How can I handle events if I need to show message only on necessary screen, for example, home screen my app", the answer is: 1) construct a campaign based on analytics event triggers you control 2) only log those analytics events on the screens you want to show related messages on

In other words, the messages show via a combination of analytics event + message to show, you're in control of both, and I think that's all there is to it.

**Reply by buraks** [reply]
has anyone managed to use in-app-messaging on react-native? My attempts today have been unsuccessful on Android/iOS both dev and release builds.
